### PR TITLE
Remove context from Backend::compile

### LINKED ISSUE #1904: Pass Context at runtime not compile time
In order to facilitate multi-threaded execution context passing needs to be moved from compile time to runtime.  **This will break the current API!**

At compile time, constant weight allocation will remain unchanged, however Placeholder memory will no longer be allocated at runtime. The expected sizes will be calculated from the the Placeholder types.

At runtime (ExecutionEngine::run) the context will be passed in and Placeholder memory will be assigned to point to the allocations in the context. 

When done context will no longer be a parameter of the compile function. ExecutionEngine::run and function::execute will expect a context parameter.
 
This change will be done in stages to minimize disruption. 

1. ExecutionEngine::run(Context &ctx) will be introduced. This is the new execution flow. When the bugs are worked out it will replace run. Tests will be ported to use run(ctx).

2. CompiledFunction::execute(Context &ctx) will be introduced, similar to run(ctx) this will replace execute() when the work is finished. Tests will be ported to execute(Context &ctx).

3. execute(Context &ctx) will be updated to handle memory allocation at runtime. 

4.  Once runctx and execute(ctx) are working and all tests pass run(ctx) will replace run and execute(ctx) will replace execute. At this point any external code will need to be updated to reflect these changes.

### PR DESCRIPTION
**Pull request by gcatron**
*Description*: Unifies backends to a single run time api and calls setup/teardown from the ExecutionEngine. Removes context from compile and compileIR as it is no longer needed at compile time.
*Testing*: ninja test, run.sh and bundles all look good.
*Documentation*: N/A
Fixes #1904

### CODE REVIEW

**Comment by rdzhabarov** [comment]
you need to rebase on latest master to pick up the docker image changes.

File: include/glow/Backends/CompiledFunction.h [hunk 1]
```diff
@@ -60,7 +60,16 @@ class CompiledFunction {
   virtual ~CompiledFunction() = default;
   /// Execute the network and allocate Placeholder memory with given
   /// \p ctx providing mapping between Placeholder and populated tensor.
-  virtual void execute(Context &ctx) = 0;
+  virtual void execute() = 0;
+
+  /// Does any needed initialization work for the Backend.
```
Comment (rdzhabarov) [inline include/glow/Backends/CompiledFunction.h]: can you add brief explanation of potential actions for each method. Copy constants, init device etc.

File: lib/ExecutionEngine/ExecutionEngine.cpp [hunk 2]
```diff
@@ -157,8 +151,8 @@ void ExecutionEngine::optimizeFunction(CompilationMode mode, Function *F) {
 void ExecutionEngine::compile(CompilationMode mode, Function *F, Context &ctx) {
   optimizeFunction(mode, F);
   // Make sure that the context has backing tensors for all placeholders.
-  ctx.allocate(M_.getPlaceholders());
-  function_ = backend_->compile(F, ctx);
+  // ctx.allocate(M_.getPlaceholders());
```
Comment (rdzhabarov) [inline lib/ExecutionEngine/ExecutionEngine.cpp]: remove commented code

**Comment by rdzhabarov** [comment]
> Finishes #1904

Do we still copy weights every run?
Anything left before enabling real multithreading for handling ONNXIFI requests?

**Comment by gcatron** [comment]
> > Finishes #1904
> 
> Do we still copy weights every run?
> Anything left before enabling real multithreading for handling ONNXIFI requests?

We do still copy weights every run. This PR moved the work up into the Execution engine layer. Next step is to design and build out the runtime engine that will handle concurrency, pipelining, and mutli-device issues. A short term solution would be to add logic to EE that only calls setupRuns on the first call of run.

**Comment by rdzhabarov** [comment]
For sanity check, once this is merged can you update fbcode accordingly.

**Pull request merged**